py4web/utils/auth_plugins/oauth2google_scoped.py

```diff
@@ -0,0 +1,199 @@
+# This is a version of google login that also enables the use of other
+# authenticatio scopes (e.g., Google Drive, etc).  The credentials for the
+# scopes are stored, so that the application can access them and use to
+# operate on the scopes (e.g., create files on Google Drive on behalf of
+# the user).
+# See https://developers.google.com/identity/protocols/oauth2/web-server#python
+
+import calendar
+import json
+import time
+import uuid
+
+import google_auth_oauthlib.flow
+import google.oauth2.credentials
+from googleapiclient.discovery import build
+
+from py4web import request, redirect, URL, HTTP
+from pydal import Field
+
+class OAuth2GoogleScoped(object):
+    """Class that enables google login via oauth2 with additional scopes.
+    The authorization info is saved so the scopes can be used later on."""
+
+    # These values are used for the plugin registration.
+    name = "oauth2googlescoped"
+    label = "Google Scoped"
+    callback_url = "auth/plugin/oauth2googlescoped/callback"
+
+    def __init__(self, secrets_file=None, scopes=None, db=None,
+                 define_tables=True, delete_credentials_on_logout=True):
+        """
+        Creates an authorization object for Google with Oauth2 and paramters.
+
+        There are some differences between this plugin and other Oauth2 plugins:
+        - The plugin uses the database, and creates an auth_credentials table to
+          store the credentials for the scopes requested.
+        - The plugin relies on some google libraries (see on top), so these
+          need to be installed.
+        - The plugin takes in input a .json credentials file that can be
+          downloaded from Google Cloud when creating the OAuth2 credentials.
+        Args:
+            secrets_file: file with secrets for Oauth2.
+            scopes: scopes desired.
+                See https://developers.google.com/drive/api/guides/api-specific-auth
+                and https://developers.google.com/identity/protocols/oauth2/scopes
+            db: Database handle.
+            define_tables: Define the tables for storing credentials?
+            delete_credentials_on_logout: if True, the credentials are cleared
+                when the user logs out.  If False, the app keeps a copy of the
+                credentials, so it can do work on behalf of the user using
+                those credentials after logout.  This can obviously
+                generate security concerns.
+        """
+
+        # Local secrets to be able to access.
+        assert secrets_file is not None, "Missing secrets file"
+        self.secrets_file = secrets_file
+        # Scopes for which we ask authorization
+        scopes = scopes or []
+        self.scopes = ["openid",
```

Comment:
It is not clear which members are supposed to be accessed publicly and which not, for the latter we should use an _ prefix